### PR TITLE
[Enhance] Refactor augmentation module

### PR DESCRIPTION
**Pull request by shijianjian**
### Description
1. Deprecated all functionals.
2. Deprecated ```return_transform``` param for augmentations.
3. Initiated ```kornia.augmentation.Sequential```.

### CODE REVIEW

File: kornia/augmentation/augmentation.py [hunk 1]
```diff
@@ -257,10 +312,15 @@ def generate_parameters(self, batch_shape: torch.Size) -> Dict[str, torch.Tensor
         return dict()
 
     def compute_transformation(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
-        return F.compute_intensity_transformation(input)
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:260]: Can we make this to happen automatically ? If the user doesn’t specify, then Internally returns identity

File: kornia/augmentation/augmentation.py [hunk 2]
```diff
@@ -396,10 +465,16 @@ def generate_parameters(self, batch_shape: torch.Size) -> Dict[str, torch.Tensor
...
 
-    def apply_transform(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
-        return F.apply_perspective(input, params, self.flags)
+    def apply_transform(
+        self, input: torch.Tensor, params: Dict[str, torch.Tensor], transform: Optional[torch.Tensor] = None
+    ) -> torch.Tensor:
+        _, _, height, width = input.shape
+        return warp_perspective(
+            input, transform, (height, width),
+            mode=self.resample.name.lower(), align_corners=self.align_corners)
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:477]: We don’t have padding mode here?

File: kornia/augmentation/augmentation.py [hunk 3]
```diff
@@ -573,7 +656,13 @@ def __init__(self, size: Union[int, Tuple[int, int]], align_corners: bool = True
         # Since PyTorch does not support ragged tensor. So cropping function happens batch-wisely.
         super(CenterCrop, self).__init__(p=1., return_transform=return_transform, same_on_batch=True, p_batch=p,
                                          keepdim=keepdim)
-        self.size = size
+        if isinstance(size, tuple):
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:659]: Can we provide an alternative here to remove Union type?

File: kornia/augmentation/augmentation.py [hunk 4]
```diff
@@ -586,21 +675,20 @@ def __repr__(self) -> str:
...
-        if isinstance(self.size, tuple):
-            size_param = (self.size[0], self.size[1])
-        elif isinstance(self.size, int):
-            size_param = (self.size, self.size)
-        else:
-            raise Exception(f"Invalid size type. Expected (int, tuple(int, int). "
-                            f"Got: {type(self.size)}.")
         return rg.center_crop_generator(
-            batch_shape[0], batch_shape[-2], batch_shape[-1], size_param, self.device)
+            batch_shape[0], batch_shape[-2], batch_shape[-1], self.size, self.device)
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:679]: I guess that’s a bug. Do we have tests?

File: kornia/augmentation/augmentation.py [hunk 5]
```diff
@@ -586,21 +675,20 @@ def __repr__(self) -> str:
...
+            params['src'].to(input), params['dst'].to(input))
+        transform = transform.expand(input.shape[0], -1, -1)
+        return transform
 
-    def apply_transform(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
-        return F.apply_crop(input, params, self.flags)
+    def apply_transform(
+        self, input: torch.Tensor, params: Dict[str, torch.Tensor], transform: Optional[torch.Tensor] = None
+    ) -> torch.Tensor:
+        return crop_by_transform_mat(
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:690]: Why don’t use directly warp_affine ?

File: kornia/augmentation/augmentation.py [hunk 6]
```diff
@@ -671,10 +759,22 @@ def generate_parameters(self, batch_shape: torch.Size) -> Dict[str, torch.Tensor
         return rg.random_rotation_generator(batch_shape[0], degrees, self.same_on_batch, self.device, self.dtype)
 
     def compute_transformation(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
-        return F.compute_rotate_tranformation(input, params)
+        angles: torch.Tensor = params["degrees"].to(input)
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:762]: Use get rotation matrix

File: kornia/augmentation/augmentation.py [hunk 7]
```diff
@@ -778,10 +878,17 @@ def precrop_padding(self, input: torch.Tensor) -> torch.Tensor:
...
+            params['src'].to(input), params['dst'].to(input))
+        transform = transform.expand(input.shape[0], -1, -1)
+        return transform
+
+    def apply_transform(
+        self, input: torch.Tensor, params: Dict[str, torch.Tensor], transform: Optional[torch.Tensor] = None
+    ) -> torch.Tensor:
+        return crop_by_transform_mat(
+            input, transform, self.size, mode=self.resample.name.lower(),
+            padding_mode='zeros', align_corners=self.align_corners)
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:891]: Use warp_affine

File: kornia/augmentation/augmentation.py [hunk 8]
```diff
@@ -868,10 +975,17 @@ def generate_parameters(self, batch_shape: torch.Size) -> Dict[str, torch.Tensor
...
+        return transform
 
-    def apply_transform(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
-        return F.apply_crop(input, params, self.flags)
+    def apply_transform(
+        self, input: torch.Tensor, params: Dict[str, torch.Tensor], transform: Optional[torch.Tensor] = None
+    ) -> torch.Tensor:
+        return crop_by_transform_mat(
+            input, transform, self.size,
+            mode=self.resample.name.lower(), padding_mode='zeros', align_corners=self.align_corners)
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:988]: Use warp_affine. Besides this function it’s extremely slow. We should allow an alternative based on indexes

File: kornia/augmentation/augmentation.py [hunk 9]
```diff
@@ -1346,13 +1486,20 @@ def __init__(self, kernel_size: Tuple[int, int],
...
             p=p, return_transform=return_transform, same_on_batch=same_on_batch, p_batch=1.)
-        self.transform = GaussianBlur2d(kernel_size, sigma, border_type)
+        self.kernel_size = kernel_size
+        self.sigma = sigma
+        self.border_type = BorderType.get(border_type)
 
     def __repr__(self) -> str:
         return self.__class__.__name__ + f"({super().__repr__()})"
 
     def generate_parameters(self, batch_shape: torch.Size) -> Dict[str, torch.Tensor]:
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:1496]: We should be able too skip implementing this from users side

File: kornia/augmentation/mix_augmentation.py [hunk 10]
```diff
@@ -99,7 +100,16 @@ def generate_parameters(self, batch_shape: torch.Size) -> Dict[str, torch.Tensor
 
     def apply_transform(self, input: torch.Tensor, label: torch.Tensor,  # type: ignore
                         params: Dict[str, torch.Tensor]) -> Tuple[torch.Tensor, torch.Tensor]:  # type: ignore
```
Comment (edgarriba) [inline kornia/augmentation/mix_augmentation.py:102]: Why there’s type ignore ?

File: kornia/augmentation/mix_augmentation.py [hunk 11]
```diff
@@ -99,7 +100,16 @@ def generate_parameters(self, batch_shape: torch.Size) -> Dict[str, torch.Tensor
...
-        return F.apply_mixup(input, label, params)
+        input_permute = input.index_select(dim=0, index=params['mixup_pairs'].to(input.device))
+        labels_permute = label.index_select(dim=0, index=params['mixup_pairs'].to(label.device))
+
+        lam = params['mixup_lambdas'].view(-1, 1, 1, 1).expand_as(input).to(label.device)
```
Comment (edgarriba) [inline kornia/augmentation/mix_augmentation.py:106]:
```suggestion
        lam = params['mixup_lambdas'].view(-1, 1, 1, 1).expand_as(input).to(label)
```

File: kornia/contrib/augmentation/sequential.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+        >>> aug_list = Sequential(
+        ...     kornia.augmentation.ColorJitter(0.1, 0.1, 0.1, 0.1, p=1.0),
+        ...     kornia.augmentation.RandomAffine(360, p=1.0),
+        ... return_transform=True)
+        >>> output, trans_mat = aug_list(input)
+        >>> trans_mat.shape
+        torch.Size([2, 3, 3])
+    """
+
+    def __init__(self, *args: _AugmentationBase, return_transform: bool = True) -> None:
```
Comment (edgarriba) [inline kornia/contrib/augmentation/sequential.py:27]: Shouldn’t we return transform all the time here? For non transform you use non.Sequential, right ?

File: kornia/geometry/transform/crop/crop2d.py [hunk 13]
```diff
@@ -237,8 +235,35 @@ def crop_by_boxes(tensor: torch.Tensor, src_box: torch.Tensor, dst_box: torch.Te
     h_out: int = int(bbox[0][0].item())
     w_out: int = int(bbox[1][0].item())
 
+    return crop_by_transform_mat(
+        tensor, dst_trans_src, (h_out, w_out),
+        mode=mode, padding_mode=padding_mode, align_corners=align_corners)
+
+
```
Comment (edgarriba) [inline kornia/geometry/transform/crop/crop2d.py:242]: I don’t see why this is needed unless we want a `crop_by_índices` but even in that case I don’t see why we don’t just directly call warp_affine

**Review by edgarriba — CHANGES_REQUESTED** [review]
Missing tests for the new sequential

File: kornia/augmentation/augmentation.py (hunk 1 above)
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:260]: Having this is to make the users aware of what to do for transformations.

File: kornia/augmentation/augmentation.py (hunk 2 above)
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:477]: It is zero padding by default. Do we need other padding modes for perspective transform?

File: kornia/augmentation/augmentation.py (hunk 3 above)
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:659]: If I remember right, the "overload" method worked for args Union types. I do not think there is a better way to replace the Union type in args.

File: kornia/augmentation/augmentation.py (hunk 5 above)
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:690]: If so, there gonna be many code duplications for all the cropping and perspective.

File: kornia/augmentation/augmentation.py (hunk 6 above)
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:762]: I do not understand what you mean here.

File: kornia/augmentation/mix_augmentation.py (hunk 10 above)
Comment (shijianjian) [reply kornia/augmentation/mix_augmentation.py:102]: It is overridding the parent class. LIke what we did for forword.

File: kornia/augmentation/augmentation.py (hunk 9 above)
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:1496]: Probably

File: kornia/augmentation/augmentation.py (hunk 8 above)
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:988]: If think a proper implementation is to detect whether this function is called under torch.no_grad() context. Then to decide the behaviour.

File: kornia/augmentation/augmentation.py (hunk 5 above)
Comment (edgarriba) [reply kornia/augmentation/augmentation.py:690]: Then add tests for that function

File: kornia/augmentation/augmentation.py (hunk 6 above)
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:756]: ```get_rotation_matrix2d``` is not pretty efficient since it will need scale tensor as well. I will keep the current implementation for now. We might update ```get_rotation_matrix2d``` later.

**Comment by shijianjian** [comment]
@edgarriba  This shall be mergeable. Then I will be working on the augmentation core.

File: kornia/augmentation/augmentation.py [hunk 14]
```diff
@@ -200,10 +235,24 @@ def generate_parameters(self, batch_shape: torch.Size) -> Dict[str, torch.Tensor
...
-        return F.compute_intensity_transformation(input)
+        return self.identity_matrix(input)
 
-    def apply_transform(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
-        return F.apply_color_jitter(input, params)
+    def apply_transform(
+        self, input: torch.Tensor, params: Dict[str, torch.Tensor], transform: Optional[torch.Tensor] = None
+    ) -> torch.Tensor:
+        transforms = [
+            lambda img: adjust_brightness(img, params['brightness_factor'].to(img.dtype) - 1),
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:244]:
```suggestion
            lambda img: adjust_brightness(img, params['brightness_factor'].to(img) - 1),
```

Comment (edgarriba) [reply kornia/augmentation/augmentation.py:244]: same for below

File: kornia/augmentation/augmentation.py [hunk 15]
```diff
@@ -396,10 +456,17 @@ def generate_parameters(self, batch_shape: torch.Size) -> Dict[str, torch.Tensor
...
+        return get_perspective_transform(
+            params['start_points'].to(input), params['end_points'].to(input))
 
-    def apply_transform(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
-        return F.apply_perspective(input, params, self.flags)
+    def apply_transform(
+        self, input: torch.Tensor, params: Dict[str, torch.Tensor], transform: Optional[torch.Tensor] = None
+    ) -> torch.Tensor:
+        _, _, height, width = input.shape
+        transform = cast(torch.Tensor, transform)
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:466]: fine for now but can be a bit annoying to cast everytime

File: kornia/augmentation/augmentation.py (hunk 6 above)
Comment (edgarriba) [reply kornia/augmentation/augmentation.py:756]: add todo note here

File: kornia/augmentation/augmentation.py [hunk 16]
```diff
@@ -778,10 +873,18 @@ def precrop_padding(self, input: torch.Tensor) -> torch.Tensor:
         return input
 
     def compute_transformation(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
-        return F.compute_crop_transformation(input, params, self.flags)
-
-    def apply_transform(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
-        return F.apply_crop(input, params, self.flags)
+        transform: torch.Tensor = get_perspective_transform(
+            params['src'].to(input), params['dst'].to(input))
+        transform = transform.expand(input.shape[0], -1, -1)
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:878]: why is this needed - no batch ?

File: kornia/augmentation/augmentation.py [hunk 17]
```diff
@@ -1292,14 +1421,13 @@ def __init__(
...
 
-    def apply_transform(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
-        return F.apply_equalize(input, params)
+    def apply_transform(
+        self, input: torch.Tensor, params: Dict[str, torch.Tensor], transform: Optional[torch.Tensor] = None
+    ) -> torch.Tensor:
+        return equalize(input)
 
 
 class GaussianBlur(AugmentationBase2D):
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:1433]: can we rename that one to `RandomGaussianBlur` ?

File: kornia/augmentation/augmentation3d.py [hunk 18]
```diff
@@ -335,10 +364,20 @@ def generate_parameters(self, batch_shape: torch.Size) -> Dict[str, torch.Tensor
             translate, scale, shear, self.same_on_batch, self.device, self.dtype)
 
     def compute_transformation(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
-        return F.compute_affine_transformation3d(input, params)
-
-    def apply_transform(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
-        return F.apply_affine3d(input, params, self.flags)
+        transform = get_affine_matrix3d(
```
Comment (edgarriba) [inline kornia/augmentation/augmentation3d.py:367]:
```suggestion
        transform: torch.Tensor = get_affine_matrix3d(
```

File: kornia/augmentation/base.py [hunk 19]
```diff
@@ -274,7 +276,7 @@ def transform_tensor(self, input: torch.Tensor) -> torch.Tensor:
 
     def identity_matrix(self, input) -> torch.Tensor:
         """Return 3x3 identity matrix."""
-        return F.compute_intensity_transformation(input)
+        return torch.eye(3, device=input.device, dtype=input.dtype).repeat(input.shape[0], 1, 1)
```
Comment (edgarriba) [inline kornia/augmentation/base.py:279]:
```suggestion
        return kornia.utils.eye_like(3, input)
```

File: kornia/augmentation/base.py [hunk 20]
```diff
@@ -316,7 +318,7 @@ def transform_tensor(self, input: torch.Tensor) -> torch.Tensor:
 
     def identity_matrix(self, input) -> torch.Tensor:
         """Return 4x4 identity matrix."""
-        return F.compute_intensity_transformation3d(input)
+        return torch.eye(4, device=input.device, dtype=input.dtype).repeat(input.shape[0], 1, 1)
```
Comment (edgarriba) [inline kornia/augmentation/base.py:321]:
```suggestion
        return kornia.utils.eye_like(4, input)
```

File: kornia/geometry/transform/projwarp.py [hunk 21]
```diff
@@ -1,5 +1,6 @@
 """Module to perform projective transformations to tensors."""
-from typing import Tuple, List
+from typing import Tuple, List, Optional
+import warnings
```
Comment (edgarriba) [inline kornia/geometry/transform/projwarp.py:4]: @shijianjian can we rename this file to a better name ? projwarp was a bad name idea after some refactors :)

**Review by edgarriba — APPROVED** [review]
LGTM . Minor comments

File: kornia/augmentation/augmentation.py (hunk 14 above)
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:244]: I removed it. It was handled inside the function. Actually, we shall refactor the kornia core to avoid using ```to``` for type conversions to let it crash.

File: kornia/augmentation/augmentation.py (hunk 16 above)
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:878]: Removed. No effect at all.

File: kornia/augmentation/augmentation.py (hunk 17 above)
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:1433]: Probably in the next version. I will add a TODO here.

File: kornia/geometry/transform/projwarp.py (hunk 21 above)
Comment (shijianjian) [reply kornia/geometry/transform/projwarp.py:4]: what do you have in mind?

File: kornia/augmentation/base.py [hunk 22]
```diff
@@ -276,7 +277,7 @@ def transform_tensor(self, input: torch.Tensor) -> torch.Tensor:
 
     def identity_matrix(self, input) -> torch.Tensor:
         """Return 3x3 identity matrix."""
-        return kornia.utils.eye_like(3, input)
+        return kornia.eye_like(3, input)
```
Comment (edgarriba) [inline kornia/augmentation/base.py:280]: we will eventually remove things from top level

Comment (shijianjian) [reply kornia/augmentation/base.py:280]: ```eye_like``` is under geometric module than utils.

Comment (edgarriba) [reply kornia/augmentation/base.py:280]: we should move then
